File: OpenSn_Logo_CAD/opensn.lua
-- Mesh
meshgen = mesh.MeshGenerator.Create({
  inputs = {
    mesh.FromFileMeshGenerator.Create({
      filename = "opensn.msh",
    }),
  },
})
grid = meshgen:Execute()

-- Material
Ng = 1
xs_source = xs.CreateSimpleOneGroup(0.1, 0.1)
xs_block = xs.CreateSimpleOneGroup(1.0e-6, 0.001)

-- Source
src_strength = {}
for g = 1, Ng do
  src_strength[g] = 0.0
end
src_strength[1] = 100.0
src = lbs.VolumetricSource.Create({ block_ids = {0}, group_strength = src_strength })

-- Quadrature
Npolar = 8
Nazimuthal = 16
pquad = aquad.CreateGLCProductQuadrature3DXYZ(Npolar, Nazimuthal)

-- Set up solver
lbs_block = {
  mesh = grid,
  num_groups = Ng,
  groupsets = {
    {
      groups_from_to = { 0, Ng -1 },
      angular_quadrature = pquad,
      angle_aggregation_type = "single",
      inner_linear_method = "petsc_gmres",
      l_abs_tol = 1.0e-6,
      l_max_its = 500,
    },
  },
  xs_map = {
    { block_ids = {0}, xs = xs_source },
    { block_ids = {1}, xs = xs_block },
  },
}

lbs_options = {
  scattering_order = 0,
  volumetric_sources = { src },
}

phys = lbs.DiscreteOrdinatesProblem.Create(lbs_block)
phys:SetOptions(lbs_options)
ss_solver = lbs.SteadyStateSolver.Create({ lbs_problem = phys })

-- Solve
ss_solver:Initialize()
ss_solver:Execute()

fflist,count = lbs.GetScalarFieldFunctionList(phys)
fieldfunc.ExportToVTKMulti(fflist,'flux')

File: OpenSn_Logo_CAD/opensn.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
OpenSn text in block example problem
"""

import os
import sys

if "opensn_console" not in globals():
    from mpi4py import MPI
    size = MPI.COMM_WORLD.size
    rank = MPI.COMM_WORLD.rank
    # Append parent directory to locate the pyopensn modules
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import MeshGenerator, FromFileMeshGenerator
    from pyopensn.xs import MultiGroupXS, CreateSimpleOneGroup
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLCProductQuadrature3DXYZ
    from pyopensn.solver import DiscreteOrdinatesProblem, SteadyStateSolver
    from pyopensn.fieldfunc import FieldFunctionGridBased

if __name__ == "__main__":

    # Mesh
    meshgen = FromFileMeshGenerator(filename="opensn.msh")
    grid = meshgen.Execute()

    # Cross sections
    Ng = 1
    xs_source = MultiGroupXS()
    xs_source.CreateSimpleOneGroup(0.1, 0.1)
    xs_block = MultiGroupXS()
    xs_block.CreateSimpleOneGroup(1.0e-6, 0.001)

    # Source
    src_strength = [0.0 for _ in range(Ng)]
    src_strength[0] = 100.0
    src = VolumetricSource(block_ids=[1], group_strength=src_strength)

    # Quadrature
    pquad = GLCProductQuadrature3DXYZ(n_polar=8, n_azimuthal=16, scattering_order=0)

    # Solver
    phys = DiscreteOrdinatesProblem(
        mesh=grid,
        num_groups=Ng,
        groupsets=[
            {
                "groups_from_to": (0, Ng - 1),
                "angular_quadrature": pquad,
                "angle_aggregation_type": "single",
                "inner_linear_method": "petsc_gmres",
                "l_abs_tol": 1.0e-6,
                "l_max_its": 300,
            }
        ],
        xs_map=[
            {"block_ids": [1], "xs": xs_source},
            {"block_ids": [2], "xs": xs_block},
        ],
        options={
            "scattering_order": 0,
            "volumetric_sources": [src],
        }
    )
    ss_solver = SteadyStateSolver(lbs_problem=phys)
    ss_solver.Initialize()
    ss_solver.Execute()

    # Export results to VTK
    fflist = phys.GetScalarFieldFunctionList()
    FieldFunctionGridBased.ExportMultipleToVTK(fflist, "flux")
